redhat-appstudio/service-provider-integration-operator
# remove k8s_token query paramter and ensure callback endpoint only takes k8s token from session

### What does this PR do?
This PR ensures that the  /authenticate endpoint of the SPI OAuth service no longer accepts k8s token from query parameter. It accepts the token only from the form body or the session, with the form body having higher priority.

Additionally, this PR ensures that the /callback endpoint accepts k8s token only from the session. This endpoint is where the service provider redirects the user after authorization, and the service provider does not and should not include any k8s token in the redirection.

### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

https://issues.redhat.com/browse/SVPI-154

### How to test this PR?

1. Deploy SPI OAuth service with an image from this branch (or quay.io/pbaran/spis:[SVPI-154](https://issues.redhat.com//browse/SVPI-154)-k8s-token ) (for testing, you should not need to set any OAuth client credentials)
2. Create any SPIAccessTokenBinding:
```shell
kubectl apply -f samples/binding-write-repo.yaml
```
3. Get the OAuth URL from binding and get your k8s token. On minikube, you can use:
```shell
OAUTH_URL=$(kubectl get spiaccesstokenbinding binding-write-repository -o=jsonpath='{.status.oAuthUrl}')
kubectl apply -f hack/give-default-sa-perms-for-accesstokens.yaml
K8S_TOKEN=$(hack/get-default-sa-token.sh)
echo "$OAUTH_URL&k8s_token=$K8S_TOKEN"
```
5. Open the URL from above in the browser and verify that you see `No active session was found. Please use /login method to authorize your request and try again. Or provide the token as a k8s_token POST form body parameter.`

6. To test that you don't get this error message using post form body parameter, you can: 
- open `hack/oauth-ui.html` page and put your OAuth URL with k8s_token and login and afterwards open the OAuth URL. This test scenario is covered by e2e test so I will not put it here.
- or use https://gist.github.com/xbaran4/011c4f2c8c1ab9396e68bd8c18e550fb and put your k8s token and OAuth URL, and open the HTML. The page will send you to the /authenticate endpoint without having to use /login endpoint.

```
quay.io/redhat-appstudio/pull-request-builds:spi-controller-#
quay.io/redhat-appstudio/pull-request-builds:spi-oauth-#
```
